internal/core/application/webhook_handler.go

```diff
@@ -0,0 +1,297 @@
+		hooks = append(hooks, hooksForAllActions...)
+	}
+	return hooks
+}
+
+// InvokeWebhooksByAction makes a POST request to every webhook endpoint
+// registered for the given action.
+// This method adopts a circuit breaker approach in order to maximize the
+// chances that every webhook gets invoked without errors.
+func (h *webhookHandler) InvokeWebhooksByAction(actionType int, payload string) error {
+	hooks := h.getHooksByAction(actionType)
+	if actionType != AllActions {
+		hooksForAllActions := h.getHooksByAction(AllActions)
+		hooks = append(hooks, hooksForAllActions...)
+	}
+
+	eg := &errgroup.Group{}
+	for i := range hooks {
+		hook := hooks[i]
+		eg.Go(func() error { return h.doRequest(hook, payload) })
+	}
+	return eg.Wait()
+}
+
+func (h *webhookHandler) addHookByAction(hook *Hook) {
+	key := []byte{byte(hook.ActionType)}
+	hooks := h.getSerializedHooks(hook.ActionType)
+	hooks = append(hooks, hook.Serialize())
+	updatedHooks := bytes.Join(hooks, separator)
+	h.store.AddToBucket(hooksByActionBucket, key, updatedHooks)
+}
+
+func (h *webhookHandler) removeHookByAction(hook *Hook) {
+	rawHooks := h.getSerializedHooks(hook.ActionType)
+
+	var index int
+	for i, buf := range rawHooks {
+		hh, _ := NewWebhookFromBytes(buf)
+		if hh.Id == hook.Id {
+			index = i
+			break
+		}
+	}
+
+	key := []byte{byte(hook.ActionType)}
+	rawHooks = append(rawHooks[:index], rawHooks[index+1:]...)
+
+	if len(rawHooks) <= 0 {
+		h.store.RemoveFromBucket(hooksByActionBucket, key)
+		return
+	}
+
+	updatedHooks := bytes.Join(rawHooks, separator)
+	h.store.AddToBucket(hooksByActionBucket, key, updatedHooks)
+}
+
+func (h *webhookHandler) getHooksByAction(actionType int) []*Hook {
+	rawHooks := h.getSerializedHooks(actionType)
+	hooks := make([]*Hook, 0, len(rawHooks))
+	for _, buf := range rawHooks {
+		hook, _ := NewWebhookFromBytes(buf)
+		hooks = append(hooks, hook)
+	}
+	return hooks
+}
+
+func (h *webhookHandler) getSerializedHooks(actionType int) [][]byte {
+	key := []byte{byte(actionType)}
+	hooks, _ := h.store.GetFromBucket(hooksByActionBucket, key)
+	if len(hooks) <= 0 {
+		return nil
+	}
+	return bytes.Split(hooks, separator)
+}
+
+func (h *webhookHandler) doRequest(hook *Hook, payload string) error {
+	_, err := h.cb.Execute(func() (interface{}, error) {
+		headers := map[string]string{
+			"Content-Type": "application/json",
+		}
+		if hook.IsSecured() {
+			token := jwt.New(jwt.SigningMethodHS256)
+			secret, _ := hex.DecodeString(hook.Secret)
+			tokenString, _ := token.SignedString(secret)
+			headers["Authorization"] = fmt.Sprintf("Bearer %s", tokenString)
+		}
+
+		status, resp, err := h.httpClient.NewHTTPRequest("POST", hook.Endpoint, payload, headers)
+		if err != nil {
+			return nil, err
+		}
+		if status != http.StatusOK {
+			return nil, fmt.Errorf(resp)
+		}
+		return nil, nil
+	})
+
+	return err
+}
+
+var webhookManager WebhookHandler
+
+func InitWebhookManager(store securestore.SecureStorage) (clean func(), err error) {
```

Comment:
We should return instance of WebhookHandler, instead of initialising var which will be referenced throughout codebase.